### PR TITLE
add method `validateAddress` to turtle-service

### PR DESCRIPTION
**Pull request by yinjun1991**
I use turtle-service in my project. And i need to check user address is valid or not, so i add a method `validateAddress` to turtle-service.
Use case:
```
curl -X POST http://localhost:8070/json_rpc -H 'Content-Type:application/json' -d '{
  "jsonrpc":"2.0",
  "id":0,
  "method":"validateAddress",
  "params":{
    "address":"TRTLxxxxxxxxxx"    // address could be a address or an integrated address
  }
}'
```

The output:
```
// invalid address
{
  "error":{
    "code":-32700,
    "data":{
      "application_code":7
    },
    "message":"Bad address"
  },
  "id":0,
  "jsonrpc":"2.0"
}

// valid
{
  "id":0,
  "jsonrpc":"2.0",
  "result":{
    "actualAddress":"",    // the actual address without paymentId
    "isIntegrated":false,   // is an integrated address
    "paymentID":""          // the paymentId extracted from integrated address, will be empty string if address is not an integrated address.
  }
}
```

### CODE REVIEW

**Comment by yin-june** [comment]
> turtle-service is in what some may call a LTS cycle meaning critical bugs or consensus rules will be updated/fixed but we are not actively adding any new features to it as it is considered deprecated in favor of wallet-api. This PR _may_ not be accepted.
> 
> Also, your IDE performed a vast number of formatting changes that need to be reverted for this to be considered. Please review the diff and narrow the diff to just the added code for further review.

But wallet-api has lots of bug haven't been fixed, It cannot be used in production.

**Comment by brandonlehmann** [comment]
> But wallet-api has lots of bug haven't been fixed, It cannot be used in production.

Deprecated & LTS does not mean removed or unsupported. This has been addressed previously. Adding new features to LTS code is not common practice nor is it taken lightly as it hinders forward movement in overall project development. 

If you are having issues with wallet-api, please report those issues (including logs) by completing the issue templates. Wallet-api is under active development and is generally stable for the majority of users. turtle-service is not under active development.

At the very least, please address the formatting issues your IDE has created in your PR and this *may* be accepted.

**Comment by yin-june** [comment]
> > But wallet-api has lots of bug haven't been fixed, It cannot be used in production.
> 
> Deprecated & LTS does not mean removed or unsupported. This has been addressed previously. Adding new features to LTS code is not common practice nor is it taken lightly as it hinders forward movement in overall project development.
> 
> If you are having issues with wallet-api, please report those issues (including logs) by completing the issue templates. Wallet-api is under active development and is generally stable for the majority of users. turtle-service is not under active development.
> 
> At the very least, please address the formatting issues your IDE has created in your PR and this _may_ be accepted.

About wallet-api, i have open two issues:
https://github.com/turtlecoin/turtlecoin/issues/952
https://github.com/turtlecoin/turtlecoin/issues/927

The bug i reported is fixed ? Wallet-api is stable ? Are you joking?

**Comment by yin-june** [comment]
> > But wallet-api has lots of bug haven't been fixed, It cannot be used in production.
> 
> Deprecated & LTS does not mean removed or unsupported. This has been addressed previously. Adding new features to LTS code is not common practice nor is it taken lightly as it hinders forward movement in overall project development.
> 
> If you are having issues with wallet-api, please report those issues (including logs) by completing the issue templates. Wallet-api is under active development and is generally stable for the majority of users. turtle-service is not under active development.
> 
> At the very least, please address the formatting issues your IDE has created in your PR and this _may_ be accepted.

I am not familiar with c++, but wallet-api is so unstable, I have to use turtle-service in my production environment. And create this PR.
If you think wallet-api is stable for now, i have no more words.

**Comment by yin-june** [comment]
> > But wallet-api has lots of bug haven't been fixed, It cannot be used in production.
> 
> Deprecated & LTS does not mean removed or unsupported. This has been addressed previously. Adding new features to LTS code is not common practice nor is it taken lightly as it hinders forward movement in overall project development.
> 
> If you are having issues with wallet-api, please report those issues (including logs) by completing the issue templates. Wallet-api is under active development and is generally stable for the majority of users. turtle-service is not under active development.
> 
> At the very least, please address the formatting issues your IDE has created in your PR and this _may_ be accepted.

I have revert the formatting, please review this PR.

File: .gitignore [hunk 1]
```diff
@@ -13,4 +13,5 @@ src/Makefile
 src/build/
 src/cmake_install.cmake
 *.swp
-.vscode
\ No newline at end of file
+.vscode
+cmake-build-debug/
```
Comment (brandonlehmann) [inline .gitignore]: This was changed by your IDE please remove this diff.

File: src/walletservice/WalletService.cpp [hunk 2]
```diff
@@ -1723,6 +1723,36 @@ namespace PaymentService
...
 
+    std::error_code WalletService::validateAddress(
+        const std::string &address,
+        bool &isIntegrated,
+        std::string &paymentID,
+        std::string &actualAddress)
+    {
+
+        if (!CryptoNote::validateAddress(address, currency)) {
+            try {
```
Comment (brandonlehmann) [inline src/walletservice/WalletService.cpp]: Leading `{` go on a new line please. Please follow formatting conventions used elsewhere in the code.  Please make sure that the correct formatting is applied throughout.

File: src/walletservice/WalletService.cpp [hunk 3]
```diff
@@ -1723,6 +1723,36 @@ namespace PaymentService
...
+                std::tie(actualAddress, paymentID) = decodeIntegratedAddress(address, currency, logger);
```
Comment (brandonlehmann) [inline src/walletservice/WalletService.cpp]: We should be able to use the below to remain consistent with code syntax used elsewhere.

```c
[actualAddress, paymentID] = decodeIntegratedAddress(address, currency, logger);
```

File: src/walletservice/WalletService.cpp [hunk 4]
```diff
@@ -1723,6 +1723,36 @@ namespace PaymentService
...
+                isIntegrated = true;
+            }
+            catch (std::system_error &x) {
+                logger(Logging::WARNING, Logging::BRIGHT_YELLOW)
```
Comment (brandonlehmann) [inline src/walletservice/WalletService.cpp]: I suggest that this is DEBUG level not warning as warning is typically used for messages that indicate a *possible* problem in the normal operation of the program/service. Considering that we expect this to throw in some cases and the method itself does not cause any changes to the wallet or the services, changing this to DEBUG makes the most sense.

File: src/walletservice/PaymentServiceJsonRpcMessages.h [hunk 5]
```diff
@@ -595,4 +595,20 @@ namespace PaymentService
         };
     };
 
+    struct ValidateAddress {
```
Comment (brandonlehmann) [inline src/walletservice/PaymentServiceJsonRpcMessages.h]: Please drop these leading `{` down to new lines

File: src/walletservice/WalletService.cpp (hunk 3 above)
Comment (yin-june) [reply src/walletservice/WalletService.cpp]: Yeah, i use the style at first, but on my mac, clion warn me 'Expected body of lambda expression', so i change to `std::tie`. I don't known how to config the idea compiler to match your environment, I'm not familiar with cpp.

Comment (yin-june) [reply src/walletservice/WalletService.cpp]: Use `[actualAddress, paymentID]` will get an error.

Comment (brandonlehmann) [reply src/walletservice/WalletService.cpp]: What error?

**Pull request closed without merging**